src/java/org/apache/cassandra/index/sai/disk/v1/sortedterms/SortedTermsWriter.java

```diff
@@ -0,0 +1,216 @@
+    // the terms except the first term in a block.
+    // The blocks should not be too large because we can't just ranfomly jump to the term inside the block,
+    // but we have to iterate through all the terms from the start of the block.
+    static final int TERMS_DICT_BLOCK_SHIFT = 4;
+    static final int TERMS_DICT_BLOCK_SIZE = 1 << TERMS_DICT_BLOCK_SHIFT;
+    static final int TERMS_DICT_BLOCK_MASK = TERMS_DICT_BLOCK_SIZE - 1;
+
+    private final IncrementalDeepTrieWriterPageAware<Long> trieWriter;
+    private final IndexOutput trieOutput;
+    private final IndexOutput termsOutput;
+    private final NumericValuesWriter offsetsWriter;
+    private final String componentName;
+    private final MetadataWriter metadataWriter;
+
+    private BytesRefBuilder prevTerm = new BytesRefBuilder();
+    private BytesRefBuilder tempTerm = new BytesRefBuilder();
+
+    private final long bytesStartFP;
+
+    private int maxLength = -1;
+    private long pointId = 0;
+
+    /**
+     * Creates a new writer.
+     * <p>
+     * It does not own the components, so you must close the components by yourself
+     * after you're done with the writer.
+     *
+     * @param componentName the component name for the SortedTermsMeta
+     * @param metadataWriter the MetadataWriter for storing the SortedTermsMeta
+     * @param termsData where to write the prefix-compressed terms data
+     * @param termsDataBlockOffsets  where to write the offsets of each block of terms data
+     * @param trieWriter where to write the trie that maps the terms to point ids
+     */
+    public SortedTermsWriter(String componentName,
+                             MetadataWriter metadataWriter,
+                             IndexOutput termsData,
+                             NumericValuesWriter termsDataBlockOffsets,
+                             IndexOutputWriter trieWriter) throws IOException
+    {
+        this.componentName = componentName;
+        this.metadataWriter = metadataWriter;
+        this.trieOutput = trieWriter;
+        SAICodecUtils.writeHeader(this.trieOutput);
+        this.trieWriter = new IncrementalDeepTrieWriterPageAware<>(trieSerializer, trieWriter.asSequentialWriter());
+        SAICodecUtils.writeHeader(termsData);
+        this.termsOutput = termsData;
+        this.bytesStartFP = termsData.getFilePointer();
+        this.offsetsWriter = termsDataBlockOffsets;
+    }
+
+    /**
+     * Appends a term at the end of the sequence.
+     * Terms must be added in lexicographic order.
+     *
+     * @throws IOException if write to disk fails
+     * @throws IllegalArgumentException if the term is not greater than the previous added term
+     */
+    public void add(final @Nonnull ByteComparable term) throws IOException
+    {
+        tempTerm.clear();
+        copyBytes(term, tempTerm);
+
+        final BytesRef termRef = tempTerm.get();
+        final BytesRef prevTermRef = this.prevTerm.get();
```

Comment:
nit: There's a lot of unnecessary `final` stuff for local variables all over, so I'll just mention it here and let it be :)